tmexp/create_bow.py

```diff
@@ -0,0 +1,218 @@
+    all_refs = sorted([ref for ref in input_dict["files_info"]])
+    doc_freq: Counter = Counter()
+    bow: Dict[str, Dict[int, Dict[str, Any]]] = {}
+    num_docwords = 0
+    num_blobs = 0
+    for file_path, blob_dict in tqdm.tqdm(input_dict["files_content"].items()):
+        refs = sorted(
+            [
+                ref
+                for ref, file_dict in input_dict["files_info"].items()
+                if file_path in file_dict
+            ]
+        )
+        lang = input_dict["files_info"][refs[0]][file_path]["language"]
+        if lang not in langs:
+            continue
+        word_counts = {}
+        for blob_hash, feature_dict in blob_dict.items():
+            word_dict: Counter = Counter()
+            for feature in features:
+                if feature not in feature_dict:
+                    continue
+                word_dict.update(feature_dict[feature])
+            if word_dict:
+                num_blobs += 1
+                doc_freq.update(word_dict.keys())
+                word_counts[blob_hash] = word_dict
+        if not word_counts:
+            continue
+        file_bows: Dict[int, Dict[str, Any]] = {}
+        prev_ref = None
+        prev_blob = None
+        if topic_model == DIFF_MODEL:
+            prev_word_counts: Counter = Counter()
+        for ref in all_refs:
+            if ref in refs:
+                cur_blob = input_dict["files_info"][ref][file_path]["blob_hash"]
+                cur_word_counts = copy.deepcopy(word_counts[cur_blob])
+            else:
+                if prev_ref is None:
+                    continue
+                cur_blob = None
+                cur_word_counts = Counter()
+            if not (cur_blob is None and topic_model == HALL_MODEL):
+                prev_id = len(file_bows) - 1
+                if prev_blob == cur_blob:
+                    file_bows[prev_id]["refs"].append(ref)
+                else:
+                    file_bows[prev_id + 1] = {"refs": [ref]}
+                    if topic_model == HALL_MODEL:
+                        file_bows[prev_id + 1]["all"] = cur_word_counts
+                    elif topic_model == DIFF_MODEL:
+                        cur_word_counts.subtract(prev_word_counts)
+                        file_bows[prev_id + 1]["added"] = {
+                            word: count
+                            for word, count in cur_word_counts.items()
+                            if count > 0
+                        }
+                        file_bows[prev_id + 1]["removed"] = {
+                            word: -count
+                            for word, count in cur_word_counts.items()
+                            if count < 0
+                        }
+                        if cur_blob is None:
+                            prev_word_counts = Counter()
+                        else:
+                            prev_word_counts = word_counts[cur_blob]
+                    num_docwords += len(
```

Comment:
Alternatively, you can remove 0 values with +cur_word_counts, as detailed in https://docs.python.org/3/library/collections.html#counter-objects before taking the length. I like the current version, just to give you some options :)

Reply:
I actually prefer using that syntax, I was looking for something like that but couldnt find it x)